pallet/staking/src/lib.rs

```diff
@@ -735,215 +827,276 @@ pub mod pallet {
+				// collators if any error occurs to prevent the chain from stalling.
+				None
+			}
+		}
+
+		/// Shift the exposure cache states.
+		///
+		/// Previous Current  Next
+		/// Next     Previous Current
+		/// Current  Next     Previous
+		///
+		/// ```nocompile
+		/// loop { mutate(2, 0, 1) }
+		/// ```
+		pub fn shift_exposure_cache_states() {
+			let (s0, s1, s2) = <ExposureCacheStates<T>>::get();
+
+			<ExposureCacheStates<T>>::put((s2, s0, s1));
 		}
 
 		/// Elect the new collators.
 		///
 		/// This should only be called by the [`pallet_session::SessionManager::new_session`].
-		pub fn elect() -> Vec<T::AccountId> {
+		pub fn elect() -> Result<Vec<T::AccountId>, DispatchError> {
+			let nominators = <Nominators<T>>::iter().collect::<Vec<_>>();
+			let ring_pool = <RingPool<T>>::get();
+			let kton_pool = <KtonPool<T>>::get();
 			let mut collators = <Collators<T>>::iter()
 				.map(|(c, cm)| {
 					let scaler = Perbill::one() - cm;
 					let mut collator_v = 0;
-					let nominators = <Nominators<T>>::iter()
+					let nominators = nominators
+						.iter()
 						.filter_map(|(n, c_)| {
-							if c_ == c {
-								let nominator_v = scaler * Self::power_of(&n);
+							if c_ == &c {
+								let nominator_v = scaler * Self::power_of(n, ring_pool, kton_pool);
 
 								collator_v += nominator_v;
 
-								Some(IndividualExposure { who: n, vote: nominator_v })
+								Some(IndividualExposure { who: n.to_owned(), vote: nominator_v })
 							} else {
 								None
 							}
 						})
 						.collect();
 
-					((c, Exposure { vote: collator_v, nominators }), collator_v)
+					((c, Exposure { commission: cm, vote: collator_v, nominators }), collator_v)
 				})
 				.collect::<Vec<_>>();
 
 			collators.sort_by(|(_, a), (_, b)| b.cmp(a));
 
+			let cache_states = <ExposureCacheStates<T>>::get();
+
+			#[allow(deprecated)]
+			call_on_exposure!(cache_states, Next::remove_all(None))?;
+
 			collators
 				.into_iter()
 				.take(<CollatorCount<T>>::get() as _)
 				.map(|((c, e), _)| {
-					<NextExposures<T>>::insert(&c, e);
-
-					c
+					call_on_exposure!(cache_states, Next::insert(&c, e))
+						.map(|_| c)
+						.map_err(Into::into)
 				})
 				.collect()
 		}
 	}
 }
 pub use pallet::*;
 
-type RewardPoint = u32;
 type Power = u32;
 type Vote = u32;
 
 type DepositId<T> = <<T as Config>::Deposit as Stake>::Item;
 
-/// On session end handler.
-///
-/// Currently it is only used to control the inflation.
-pub trait OnSessionEnd<T>
+/// Inflation and reward manager.
+pub trait InflationManager<T>
 where
 	T: Config,
 {
 	/// Generic session termination procedures.
 	fn on_session_end() {
 		let inflation = Self::inflate();
 		let reward = Self::calculate_reward(inflation);
-		let actual_payout = <Pallet<T>>::payout(reward);
+		let actual_reward = <Pallet<T>>::dispatch_session_reward(reward);
 
-		Self::clean(inflation.unwrap_or(reward).saturating_sub(actual_payout));
+		if inflation != 0 {
```

Comment:
Suggest changing the condition to `if (inflation - actual_reward) > 0`. Much easier to read and if these two values are equal, no clear happens.